tests/virt_autotest/xen_guest_irqbalance.pm

```diff
@@ -0,0 +1,144 @@
+        foreach my $irq_id (@nic_irqs_id) {
+            assert_script_run("ssh root\@$guest \"echo 1 > /proc/irq/$irq_id/smp_affinity\"");
+            $cat_affinity_files_cmdline .= " $irq_id/smp_affinity";
+        }
+        $cat_affinity_files_cmdline .= "\"";
+        my @affinity_after_binding = split('\n', script_output($cat_affinity_files_cmdline, 60));
+        foreach my $affinity (@affinity_after_binding) {
+            record_soft_failure("NIC IRQs were binded to one cpu core successfully, which is not expected.") if $affinity eq 1;
+        }
+
+        # Generate xen guest network load
+        #julie: need find a proper download
+        #julie: 4 sec to download this TW en.tlk in OSD
+        my $download_stuff = "http://download.opensuse.org/distribution/openSUSE-current/repo/oss/boot/x86_64/loader/en.tlk";
+        assert_script_run("ssh root\@$guest \"wget $download_stuff\"");
+
+        # re-caculate the network interrupts distribution
+        # the increased NIC IRQs should be distributed in CPU cores in balance
+        my @total_irqs_on_cpu_after_network_download = get_nic_irqs_distribution_from_guest($guest, $nproc);
+        my @increased_irqs_on_cpu;
+        for (my $cpu_id = 0; $cpu_id < $nproc; $cpu_id++) {
+            $increased_irqs_on_cpu[$cpu_id] = $total_irqs_on_cpu_after_network_download[$cpu_id] - $initial_total_irqs_on_cpu[$cpu_id];
+            die("The value of network interrupts from download for CPU" . $cpu_id . " is 0") if $increased_irqs_on_cpu[$cpu_id] == 0;
+        }
+        record_info("IRQs of downloads distribution", "@increased_irqs_on_cpu");
+    }
+
+}
+
+#set up guest test environment to run irqbalance test
+sub prepare_guest_for_irqbalance {
+    my $vm_name = shift;
+
+    #4 or more vcpu is needed
+    script_run("virsh dumpxml $vm_name | grep vcpu");  #julie debug
+    my $nproc = script_output "virsh vcpucount $vm_name | grep current | grep config | awk '{ print \$3 }'";
+    if ($nproc < 4) {
+        assert_script_run "virsh dumpxml --inactive $vm_name > $vm_name.xml";
+        script_run "virsh destroy $vm_name";
+        assert_script_run "xmlstarlet edit -L -d /domain/vcpu $vm_name.xml";
+        assert_script_run "xmlstarlet edit -L \\
+                               -s /domain -t elem -n vcpu -v 4 \\
+			       -s /domain/vcpu -t attr -n placement -v static \\
+			       $vm_name.xml";
+        script_run "virsh undefine $vm_name" unless (script_run "virsh undefine --nvram $vm_name") == 0;
+        assert_script_run " ! virsh list --all | grep $vm_name";
+        assert_script_run "virsh define $vm_name.xml";
+        assert_script_run "virsh start $vm_name";
+    }
+    script_run("virsh dumpxml $vm_name | grep vcpu");  #julie debug
+
+    wait_guest_online($vm_name);
+    unless (script_run("ssh root\@$vm_name \"systemctl status irqbalance\"") == 0) {  #julie: if not running, rc still be 0?
+        assert_script_run "ssh root\@$vm_name \"zypper in irqbalance\"";
```

Comment:
Would better to add -y option. 
`assert_script_run "ssh root\@$vm_name \"zypper in -y irqbalance\""; `